---
title: "Practicals GWAS. Exploring genotype data and QC"
author: "Sodbo Sharapov"
date: "January 22, 2019"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# Outline of the practicals

The practical session consists of three steps:

 1. Exploring phenotype data
 2. Exploring genotype data and quality control
 3. Association analysis and GWAS

## Loading genotype data

Load GenABEL package
 
```{r}
library(GenABEL)
```

Load phenotype and genotype data

```{r}

data("ge03d2")


```

Let's discover the structure of ge03d2 object

```{r}
str(ge03d2)
```

The other part of an object of gwaa.data-class is gtdata, which contains
all GWA genetic information in an object of class snp.data class. It is not
supposed to be modified directly by user. Genotypes are stored as a huge 
N x M table, where N is number of IDs and M is a number of SNPs.
The genotypic data can be accessed through the gtdata function, e.g.

```{r}
gtdata(ge03d2[1:10,1:10])
```

gtdata can be transformed to numeric or character data
```{r}

as.numeric(gtdata(ge03d2[1:5,1:5]))

as.character(gtdata(ge03d2[1:5,1:5]))
```

The number of individuals can be obtained by "nids" function; and the number of SNPs - by "nsnps" function:

```{r}
nids(ge03d2)
nsnps(ge03d2)
```

Which SNP information do we have?

```{r}
snp_info <- summary(gtdata(ge03d2))

head(snp_info)
```

Description of some columns:
  
  1. Q.2 - frequency of allele 2 (A2)
  2. Pexact - p-value of HWE test
  3. P.11, P.12, P.22 - number of genotypes

What is the distribution of number of SNPs per chromosome?

```{r}

table(chromosome(ge03d2))
```

## Summary for a single SNP

Let's extract information about particular SNP

```{r}

snp_info['rs6212914',]
```

What is the distribution of genotypes for this SNP across samples

```{r}

table(as.character(gtdata(ge03d2[,'rs6212914'])))

```

Let's estiamte effective allele frequency for this SNP by different ways

```{r}

(snp_info['rs6212914','P.12'] + snp_info['rs6212914','P.22'] * 2) / snp_info['rs6212914','NoMeasured'] / 2

mean(as.numeric(gtdata(ge03d2[,'rs6212914'])), na.rm=TRUE) / 2

```
Let's perform HWE test for this SNP

```{r}
HWE.show(ge03d2[,'rs6212914'])
```

## Per-SNP Summary of genotype data

Let's have a look at the summary of the genotype data.

First, let's add EAF (effective allele frequency) column to the *snp_info* object. Note, that EAF is also presented in 'Q.2' column, but we will calculate it by ourself.

```{r}

snp_info$EAF <- (snp_info[,'P.12'] + snp_info[,'P.22'] * 2) / snp_info[,'NoMeasured'] / 2

head(snp_info)

```

Let's have a look at the distribution of EAF.

```{r}
hist(snp_info$EAF, breaks=50)
```

There are a lot of SNPs with EAF below 0.05 and above 0.95.

What about call rate?

```{r}
hist(snp_info$CallRate, breaks=50)
```

What about HWE test?

```{r}

catable(snp_info$Pexact, c(0.05/nsnps(ge03d2), 0.01, 0.05, 0.1), cum=TRUE)
```

What if we will test HWE using controls?

```{r}

control_ids <- idnames(ge03d2)[phdata(ge03d2)$dm2==0]

catable(summary(gtdata(ge03d2[control_ids,]))$Pexact, c(0.05/nsnps(ge03d2), 0.01, 0.05, 0.1), cum=TRUE)
```
As you can see, now much less SNPs show deviation from HWE.

## Per-ID Summary of genotype data

Now let's check the per-ID call rate and heterozygosity rate.

First let's get per-ID summary of the data.
```{r}
idsummary <- perid.summary(ge03d2)

head(idsummary)
```

Description of some columns:

  1. NoMeasured - number of measured SNPs
  2. NoPoly - number of polymorfic SNPs (SNPs with minor allele frequency > 0)
  3. Het - heterozigosity rate
  
Let's check the distribution of per-ID call rate. It can give an information about samples with low callrate, that may indicate problems with these samples.

```{r}

hist(idsummary$CallPP, breaks=100)

```

Few samples have low per-ID call rate.

What about heterozigosity rate?

```{r}

hist(idsummary$Het, breaks=100)

```

Few samples have outlying heterozigosity rate, that may indicate DNA sample contamination.

## Summary of exploring genotype data

Let's see the descriptive statistics for all SNP markes in our data.

```{r}
descriptives.marker(ge03d2)
descriptives.marker(ge03d2, ids=control_ids)
```


## Quality control of genotype data

Now let's summarize our observations and perform quality control for genotype data.

Basic QC consists of following steps:

  1. Identification of individuals with discordant sex information
    1. X-Heterozygose males
    2. Y-Heterozygose females
  
  2. Identification of individuals with elevated missing data rates or outlying heterozygosity rate
    1. IDs with low call rate (<95%, 97% or 99%)
    2. IDs with outlying heterozygosity rate
      1. Too high - DNA sample contamination 
      2. Too low - inbreeding
  
  3. Identification of duplicated or related individuals (if these individuals need to be removed)
    1. Estimation of kinship matrix
    2. IBS=0.5 -> monozygos twins or duplicated samples
    3. IBS > 0.125 -> close relatives
  
  4. Identification of individuals of divergent ancestry
    1. PCA analysis using kinship matrix
    2. 1000 Genomes project data can be used as reference sample
  
  5. Removal of all individuals failing QC
  
  6. Identification of all markers with an excessive missing data rate
    1. SNP call rate below 97%
  
  7. Removal of all markers failing QC
  

Let's perform QC using checkmarker function.

```{r}

qc1 <- check.marker(ge03d2, p.level = 0, maf = 0.01, perid.call = 0.97, callrate = 0.97)

```

Now let's define IDs which will be used for HWE test.

```{r}
hweids <- intersect(control_ids, qc1$idok)

qc2 <- check.marker(ge03d2, 
                    p.level = 1e-6, 
                    maf = 0.01, 
                    perid.call = 0.97, 
                    callrate = 0.97,
                    hweidsubset = hweids)
```

QC is done. Now let's save QCed data

```{r}

ge03d2clean <- ge03d2[qc2$idok,qc2$snpok]

```

## Finding genetic sub-structure

Now, we are ready for the second round of QC – detection of genetic outliers
which may contaminate our results. We will detect genetic outliers using a
technique suggested by Price at al.
As the first step, we will compute a matrix of genomic kinship between all
pairs of individuals, using autosomal markers:

```{r}
gkin <- ibs(ge03d2clean[, autosomal(ge03d2clean)], weight="freq")

gkin[1:5, 1:5]
```

The numbers below the diagonal show the genomic estimate of kinship (aka
’genomic kinship’ or ’genome-wide IBS’), the numbers on the diagonal corre-
spond to 0.5 plus the genomic homozigosity, and the numbers above the diagonal
tell how many SNPs were typed successfully for both subjects (thus the IBD
estimate is derived using this number of SNPs).
Second, we transform this matrix to a distance matrix using standard R
command

```{r}
data.dist <- as.dist(0.5 - gkin)
```

Finally, we perform Classical Multidimensional Scaling by
```{r}
data.mds <- cmdscale(data.dist)
```

We can present the results graphically by
```{r}
plot(data.mds)
```

Each point on the plot corresponds to a person, and the 2D distances between 
points are fitted to be as close as possible to those presented in the original distance matrix.
You can see that study participants clearly cluster in two groups. 
Let's get id names of samples from the second cluster and removed them from the data and re-run QC

```{r}

clean_ids <- rownames(data.mds)[data.mds[,1]<0.1]

controls_clean <- intersect(clean_ids,idnames(ge03d2clean)[phdata(ge03d2clean)$dm2==0])

qc3 <- check.marker(ge03d2[clean_ids,], 
                    p.level = 1e-6, 
                    maf = 0.01, 
                    perid.call = 0.97, 
                    callrate = 0.97,
                    hweidsubset = controls_clean)

ge03d2clean <- ge03d2[qc3$idok,qc3$snpok]
```

Finally, let's save the data for GWAS

```{r}
save(ge03d2clean, file='/home/sodbo/work/2019_winter_school/src/data/ge03d2clean.RData')
```
## Additional literature: 

Aulchenko, Yurii S., Karssen, Lennart C., & The GenABEL project developers. (2015). The GenABEL Tutorial. Zenodo. http://doi.org/10.5281/zenodo.19738

Anderson, C. A., Pettersson, F. H., Clarke, G. M., Cardon, L. R., Morris, A. P., & Zondervan, K. T. (2010). Data quality control in genetic case-control association studies. Nature Protocols, 5(9), 1564–73. http://doi.org/10.1038/nprot.2010.116
## Saving data
