---
title: "Practicals GWAS. Exploring phenotype data."
author: "Sodbo Sharapov"
date: "January 22, 2019"
output:
  html_document: 
    keep_md: true
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# Outline of the practicals

The practical session consists of three steps:

 1. Exploring phenotype data
 2. Exploring genotype data and quality control
 4. Association analysis and GWAS

## Loading phenotype data

Load GenABEL package
 
```{r}
library(GenABEL)
```

Load phenotype and genotype data

```{r}

data("ge03d2")


```

Let's get the Structure of ge03d2 object

```{r}
str(ge03d2)
```

The object you have loaded, ge03d2, belongs to the gwaa.data class. This is a
special R class developed to facilitate GWA analysis.
In GWA analysis, different types of data are used. These include the pheno-
typic and genotypic data on the study participants and chromosome and location
of every SNP. For every SNP, it is desirable to know the details of coding (how
are the alleles coded? – A, T, G, C? – as well as the strand – ’+’ or ’-’, ’top’ or
’bot’ ? – this coding is for).
One could attempt to store all phenotypes and genotypes together in a single
table, using, e.g. one row per study subject; than the columns will correspond
to study phenotypes and SNPs. For a typical GWA data set, this would lead
to a table of few thousands rows and few hundreds of thousands to millions of
columns. Such a format is generated when one downloads HapMap data for a
region. To store GWA data in such tables internally, within R, proves to be
inefficient. In GenABEL-package, a special data class, gwaa.data-class is used
to store GWA data.
You may consider an object of gwaa.data-class as a ’black box’ from which
you can get specific data using specific functions. If you are interested in the in-
ternal structure of the gwaa.data-class, you can find the description in section
B.1 (Internal structure of gwaa.data-class) in GenABEL tutorial.

The number of individuals described in an object of gwaa.data-class can
be accessed through nids function

```{r}
nids(ge03d2)
```

Let's save phenotype data into separate object named *pheno*.
And look at the first rows of the *pheno* object

```{r}

pheno <- phdata(ge03d2)

head(pheno)
```
The rows of this data frame correspond to the study participants (individuals), and the columns
correspond to the variables (phenotypes). There are two default variables, which are always
present in phdata. The first of these is “id”, which contains participant's
identification code. This identification code can be arbitrary character, number,
or alphanumeric combination, but every person must be coded with an unique
ID. The second default variable is “sex”, where males are coded with ones (“1”)
and females are coded with zero (“0”). Sometimes gender can be coded as "1" for males and "2" for females.

## Summary of phenotype data

Description of the columns:

  1. ID: id code of sample 
  2. sex: sex, '0' is female, '1' is male
  3. age: age, in years
  4. dm2: diabetes mellitus type 2, '0' is unaffected, '1' is affected
  5. height: height, in cm
  6. wieght: wieght, in kg
  7. diet: following the diet, '0' is false, '1' is true
  8. bmi: BMI, body mass index, in $cm/kg^2$

Now let's have a look at the summary of the phenotype data

```{r}
summary(pheno)
```

For the categorial traits **table** function can be applied to get summary of the trait.

```{r}
table(pheno$sex)
table(pheno$dm2)
table(pheno$diet)
```

## Distribution of traits

You can plot a historgram for quantitaive traits.

```{r}
hist(pheno$age, main='Histogram of age', xlab = 'Age')
hist(pheno$height, main='Histogram of height', xlab = 'Height')
hist(pheno$weight, main='Histogram of weight', xlab = 'Weight')
hist(pheno$bmi, main='Histogram of bmi', xlab = 'BMI')
```

## Test for normality

You can test whether the trait's discribution is normal or not using Shapiro test.

```{r}
shapiro.test(pheno$height)
shapiro.test(pheno$weight)
shapiro.test(pheno$bmi)
```

Distribution of BMI has left skewness. Log transformation can help to transform it to normal.

```{r}
hist(log(pheno$bmi), main='Histogram of logBMI', xlab = 'logBMI')
shapiro.test(log(pheno$bmi))
```



## Correlation structure

It is important to check whether your trait of interest is correlated with covariates.
To eheck this you can estimate the Peasron correlation coefficient and/or perform linear regression analysis.
First, let's examine the correlation structure of phenotype data.

```{r}
cor_matrix <- cor(pheno[,c('dm2','sex','age','height','weight','bmi','diet')], use='complete.obs')

cor_matrix
```
Now let's vizualize correlation matrix using *corplot* package

```{r}
library(corrplot)

corrplot(cor_matrix)
```

Heatmap clearly shows that dm2 is correlated with weight and bmi.

## Logistic regression of outcome onto covariates

Let's run logistic regression for dm2 vs weight

```{r}
dm2_weight <- glm(dm2~weight,data = pheno,family = 'binomial')
summary(dm2_weight)
```


```{r}
dm2_bmi <- glm(dm2~bmi,data = pheno,family = 'binomial')
summary(dm2_bmi)
```

```{r}
dm2_sex <- glm(dm2~sex,data = pheno,family = 'binomial')
summary(dm2_sex)
```

betaOR for sex is 0.4465, that corresponds to `OR=exp(0.4465)=1.563`. It is concordant with that, calculated from 2x2 table:
```{r}
table(pheno$dm2,pheno$sex)
(275*253)/(210*212)
```
