---
title: "Practicals GWAS. Association analysis and GWAS"
author: "Sodbo Sharapov"
date: "January 22, 2019"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# Outline of the practicals

The practical session consists of three steps:

 1. Exploring phenotype data
 2. Exploring genotype data and quality control
 3. Association analysis and GWAS

## Loading QCed ge03d2clean data

Load GenABEL package
 
```{r}
library(GenABEL)
```

Load clean phenotype and genotype data

```{r}

load('/home/sodbo/work/2019_winter_school/src/data/ge03d2clean.RData')


```

Let's check the number of IDs and SNPs

```{r}
nids(ge03d2clean)

nsnps(ge03d2clean)
```

## Single SNP association analysis

Let's run association analysis for a single SNP.
For this we will store genotypes for this SNP in a separate object.

```{r}

snpnames(ge03d2clean)[1:10]

my_snp_num <- as.numeric(gtdata(ge03d2clean[,'rs1646456']))

my_snp_cha <- as.character(gtdata(ge03d2clean[,'rs1646456']))

```

Let's check the distribution of genotypes across samples

```{r}

table(my_snp_cha)
table(my_snp_num)
```

Let's run logistic regresion with SNP as predictor and dm2 as outcome.

```{r}
log_reg_snp <- glm(phdata(ge03d2clean)$dm2~my_snp_num)

summary(log_reg_snp)
```

As we can see, there is no association (p-value>0.05).
Let's check the distribution of cases and controls across the genotypes.

```{r}
table(phdata(ge03d2clean)$dm2,my_snp_cha)
```

Indeed, the distribution of cases and controls across the genotypes is random.

## Genome-wide association scan

Now let's run your first small GWAS :). We will use 'mlreg' function of GenABEL package.
Basically, we will run 7000 regressions. We will run GWAS for dm2 with age and sex as covariates

```{r}
qt_simple <- mlreg(dm2~sex+age, data=ge03d2clean, trait="binomial")
```

Let's have a look at the top 10 signals

```{r}
descriptives.scan(qt_simple , sort="Pc1df")
```

As you can see, the top associated SNP has P-value of association of 3e-7,
which is not significant at the genome-wide significance level of 5e-8.

But what would happend if we will bmi as covariate into the model?

```{r}
qt_sex_age_bmi <- mlreg(dm2 ~ sex + age + bmi, ge03d2clean, trait="binomial")

descriptives.scan(qt_sex_age_bmi , sort="Pc1df")

```

Now we see one SNP with P-value of 1.056e-08, that is genome-wide significantly associated with T2D.

## Manhattan plot

Let's plot the Manhattan plot
```{r}
plot(qt_sex_age_bmi,df = 1)
```

## Regional assoc plot

Now let's zoom Manhattan plot around found signal. This plot is called regional association plot.
```{r}
gwas_sum <- summary(qt_sex_age_bmi)

gwas_sum_sm <- gwas_sum[gwas_sum$Chromosome==1 & gwas_sum$Position>1047389-250000 & gwas_sum$Position<1047389+250000,]

plot(-log10(gwas_sum_sm$P1df)~gwas_sum_sm$Position)

```

## Table of genotypes and c/c status for the best SNP

Now let's have a look at the distribution of cases and controls across the genotypes of the
top associated SNP.

```{r}
table(phdata(ge03d2clean)$dm2,as.character(gtdata(ge03d2clean[,'rs7903146'])))
```

As you can see, the effect size is not big. Let's estimate OR for allele T.

```{r}
exp(0.7898)
```

Let's compare results of mlreg with glm

```{r}
summary(glm(dm2~sex+age+bmi+as.numeric(gtdata(ge03d2clean[,'rs7903146'])),family='binomial', data=phdata(ge03d2clean)))
```

As you can see, the results between mlreg and glm are concordant.

## Saving summary stats

```{r}
sum_stats <- results(qt_sex_age_bmi)

snp_info <- summary(gtdata(ge03d2clean))

sum_stats2write <- sum_stats[,c('Chromosome','Position','A2','A1','N','effB','se_effB','P1df','Strand')]

sum_stats2write <- cbind(rownames(snp_info),rownames(snp_info),sum_stats2write)

sum_stats2write <- cbind(sum_stats2write[,c(1:7)],snp_info$Q.2,sum_stats2write[,c(8:11)])

colnames(sum_stats2write) <- c('rsid','snpid','chr','pos','a1','a0','n','freq1','beta1','se','p','strand')

write.table(sum_stats2write,file='/home/sodbo/work/2019_winter_school/src/results/sum_stats.tsv', sep='\t', quote = FALSE,row.names = FALSE)
```
## Additional literature: 

Aulchenko, Yurii S., Karssen, Lennart C., & The GenABEL project developers. (2015). The GenABEL Tutorial. Zenodo. http://doi.org/10.5281/zenodo.19738

Anderson, C. A., Pettersson, F. H., Clarke, G. M., Cardon, L. R., Morris, A. P., & Zondervan, K. T. (2010). Data quality control in genetic case-control association studies. Nature Protocols, 5(9), 1564â€“73. http://doi.org/10.1038/nprot.2010.116

## Saving data
