---
title: "Practicals GWAS. Exploring genotype data"
author: "Sodbo Sharapov"
date: "January 22, 2019"
output:
  html_document: 
    keep_md: true
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# Outline of the practicals

The practical session consists of five steps:

 1. Exploring phenotype data
 2. Exploring genotype data
 3. Quality control of genotype data
 4. Association analysis and GWAS
 5. Vizualization of the results

## Loading genotype data

Load GenABEL package
 
```{r}
library(GenABEL)
```

Load phenotype and genotype data

```{r}

data("ge03d2")


```

Let's get the Structure of ge03d2 object

```{r}
str(ge03d2)
```

Genotypes are stored as a huge N x M table, where N is number of IDs and M is a number of SNPs.

```{r}

as.numeric(gtdata(ge03d2[1:5,1:5]))

as.character(gtdata(ge03d2[1:5,1:5]))
```

How many IDs and SNPs do we have in our data?

```{r}
nids(ge03d2)
nsnps(ge03d2)
```

What information about SNPs do we have?

```{r}
snp_info <- summary(gtdata(ge03d2))

head(snp_info)
```

Description of columns:
  
  1. ...
  2. ...
  3. ...

What is the distribution of number of SNPs per chromosome?
```{r}

table(chromosome(ge03d2))
```

## Summary for a single SNP

Let's extract information about particular SNP

```{r}

snp_info['rs6212914',]
```

What is the distribution of genotypes for this SNP across samples

```{r}

table(as.character(gtdata(ge03d2[,'rs6212914'])))

```

Let's estiamte effective allele frequency for this SNP using different ways

```{r}

(snp_info['rs6212914','P.12'] + snp_info['rs6212914','P.22'] * 2) / snp_info['rs6212914','NoMeasured'] / 2

mean(as.numeric(gtdata(ge03d2[,'rs6212914'])), na.rm=TRUE) / 2

```
Let's perform HWE test for this SNP

```{r}
HWE.show(ge03d2[,'rs6212914'])
```

## Per-SNP Summary of genotype data

Let's have a look at the summary of the genotype data.

First, let's add EAF (effective allele frequency) column to the *snp_info* object

```{r}

snp_info$EAF <- (snp_info[,'P.12'] + snp_info[,'P.22'] * 2) / snp_info[,'NoMeasured'] / 2

head(snp_info)

```

Let's look at the distribution of EAF - effective allele frequency.

```{r}
hist(snp_info$EAF, breaks=50)
```

There are a lot of SNPs with EAF below 0.05 and above 0.95.

What about call rate?

```{r}
hist(snp_info$CallRate, breaks=50)
```

What about HWE test?

```{r}

catable(snp_info$Pexact, c(0.05/nsnps(ge03d2), 0.01, 0.05, 0.1), cum=TRUE)
```

What if we will test HWE using controls?

```{r}

control_ids <- phdata(ge03d2)$dm2==0

catable(summary(gtdata(ge03d2[control_ids,]))$Pexact, c(0.05/nsnps(ge03d2), 0.01, 0.05, 0.1), cum=TRUE)
```
As you can see, now much less SNPs show deviation from HWE.

## Per-ID Summary of genotype data

Now let's check the per-ID SNP call rate and heterozygosity rate.

First let's get per-ID summary of the data.
```{r}
idsummary <- perid.summary(ge03d2)

head(idsummary)
```

Description of columns:

  1. ...
  2. ...
  3. ...
  
Let's check the distribution of per-ID call rate. It can give information about samples with low callrate, that may indicate problems with these samples.

```{r}

hist(idsummary$CallPP, breaks=100)

```

Few samples have low per-ID call rate.

What about heterozigosity rate?

```{r}

hist(idsummary$Het, breaks=100)

```

Few samples have outlying heterozigosity rate, that may indicate DNA sample contamination.

## Summary of exploring genotype data

Let's see the descriptive statistics for all SNP markes in our data.

```{r}
descriptives.marker(ge03d2)
descriptives.marker(ge03d2, ids=control_ids)
```


## Quality contron of genotype data

Now let's summarize our observations and perform quality control for genotype data.

Basic QC consists of following steps;

  1. Remomving SNPs with
    1. 
    2. 
  
  2. Removing IDs with
    1.
    2.
    
  3. Run again steps (1) and (2)
  
  4. Define population sub-structure
  
  5. Remove IDs with outlying ancestry.


Additional literature: 

Aulchenko, Yurii S., Karssen, Lennart C., & The GenABEL project developers. (2015). The GenABEL Tutorial. Zenodo. http://doi.org/10.5281/zenodo.19738

Anderson, C. A., Pettersson, F. H., Clarke, G. M., Cardon, L. R., Morris, A. P., & Zondervan, K. T. (2010). Data quality control in genetic case-control association studies. Nature Protocols, 5(9), 1564â€“73. http://doi.org/10.1038/nprot.2010.116
## Saving data